contract/r/gnoswap/gov/staker/v1/state.gno

```diff
@@ -159,65 +161,124 @@ func (g *govStakerV1) getUserDelegationsWithDelegatee(user address, delegatee ad
+//   - timestamp: timestamp of the delegation change
+func (g *govStakerV1) addDelegationRecord(delegateeAddr address, amount int64, timestamp int64) {
+	// Update total delegation history
+	g.updateTotalDelegationHistory(amount, timestamp)
+
+	// Update user delegation history
+	g.updateUserDelegationHistory(delegateeAddr, amount, timestamp)
 }
 
-// addDelegationRecord adds a new delegation record to history and updates snapshots.
+// updateTotalDelegationHistory updates the total delegation history with cumulative value.
 //
 // Parameters:
-//   - delegationRecord: delegation record to add
-func (g *govStakerV1) addDelegationRecord(delegationRecord *staker.DelegationRecord) {
-	// Add to history
-	if err := g.store.AddDelegationRecord(delegationRecord); err != nil {
-		panic(err)
+//   - amount: amount change (positive for delegate, negative for undelegate)
+//   - timestamp: timestamp of the change
+func (g *govStakerV1) updateTotalDelegationHistory(amount int64, timestamp int64) {
+	history := g.store.GetTotalDelegationHistory()
+
+	// Get current total from the most recent entry
+	currentTotal := g.getLatestTotalDelegation(history)
+	newTotal := safeAddInt64(currentTotal, amount)
+
+	if newTotal < 0 {
+		newTotal = 0
 	}
 
-	snapshot := g.store.GetDelegationSnapshot()
-	resolver := NewDelegationSnapshotResolver(snapshot)
-	snapshot = resolver.AddRecord(delegationRecord)
-	if err := g.store.SetDelegationSnapshot(snapshot); err != nil {
+	history.Set(timestamp, newTotal)
+
+	if err := g.store.SetTotalDelegationHistory(history); err != nil {
 		panic(err)
 	}
 }
 
-// setDelegationHistory replaces the current delegation history.
+// updateUserDelegationHistory updates the user delegation history with cumulative values.
+// Structure: address -> *UintTree[timestamp -> int64]
 //
 // Parameters:
-//   - history: new delegation history to set
-func (g *govStakerV1) setDelegationHistory(history staker.DelegationHistory) {
-	if err := g.store.SetDelegationHistory(history); err != nil {
+//   - delegateeAddr: address of the delegatee
+//   - amount: amount change (positive for delegate, negative for undelegate)
+//   - timestamp: timestamp of the change
+func (g *govStakerV1) updateUserDelegationHistory(delegateeAddr address, amount int64, timestamp int64) {
+	history := g.store.GetUserDelegationHistory()
+	addrStr := delegateeAddr.String()
+
+	// Get or create user's timestamp tree
+	var userHistory *staker.UintTree
+
+	if existingHistory, exists := history.Get(addrStr); exists {
+		userHistoryInt, ok := existingHistory.(*staker.UintTree)
+		if !ok {
+			panic(ufmt.Sprintf("invalid user history type: %T", existingHistory))
+		}
+
+		userHistory = userHistoryInt
+	} else {
+		userHistory = staker.NewUintTree()
+	}
+
+	// Get current amount from the most recent entry
+	currentAmount := g.getLatestUserDelegation(userHistory)
+	newAmount := safeAddInt64(currentAmount, amount)
```

Comment:
This code allows negative delegation values from users. It seems best to handle it with a clamp, like Total Delegation.

```go
if newTotal < 0 {
	newTotal = 0
}
```